jax/random.py

```diff
@@ -876,20 +880,43 @@ def _case4(zagf):
     return grad
 
 def _gamma_grad(sample, a):
-    samples = np.reshape(sample, -1)
-    alphas = np.reshape(a, -1)
+  samples = np.reshape(sample, -1)
+  alphas = np.reshape(a, -1)
+  if xla_bridge.get_backend().platform == 'cpu':
+    grads = lax.map(lambda args: _gamma_grad_one(*args), (samples, alphas))
+  else:
     grads = vmap(_gamma_grad_one)(samples, alphas)
-    return grads.reshape(onp.shape(a))
+  return grads.reshape(onp.shape(a))
 
-@custom_transforms
 def _gamma_impl(key, a):
-    alphas = np.reshape(a, -1)
-    keys = split(key, onp.size(alphas))
+  a_shape = np.shape(a)
+  # split key to match the shape of a
+  key_ndim = np.ndim(key) - 1
+  key = np.reshape(key, (-1, 2))
+  key = vmap(split, in_axes=(0, None))(key, prod(a_shape[key_ndim:]))
```

Comment:
This logic, which treats `key` like an array, is a bit of an abstraction violation.

Can this be written more like `keys = split(key, prod(a_shape))`, where we assume that `key.shape == (2,)` for the key passed into `gamma`?

Reply:
Ah, I see: it's because the batching rule `_gamma_batching_rule` might set up leading batch dimensions.

Nevermind the above comment, this seems good, at least until we think of something better (which is true of all code: good until we think of something even better!).